README
README file for CGI::Snapp.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip CGI-Snapp-1.00.tgz
shell>tar mxvf CGI-Snapp-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Snapp.pm through you favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip CGI-Snapp-1.00.zip
shell>ppm install --location=. CGI-Snapp
shell>del CGI-Snapp-1.00.ppd
shell>del PPM-CGI-Snapp-1.00.tar.gz


CHANGES
Revision history for Perl extension CGI::Snapp.

1.05  Thu Apr 18 13:08:00 2013
	- Unreleased.
	- Add a test of add_header() to t/headers.pl.

1.04  Fri Dec 14 16:38:00 2012
	- Unreleased
	- Change sub log():
		From: $myself -> logger -> $level($s)        if ($myself && $myself -> logger)
		To:   $myself -> logger -> log($level => $s) if ($myself && $myself -> logger)
		The reason for this is that if your logger is derived from another class or role, that class
		or role had to provide not just the log() method, but all of debug(), info(), notice(), warning(),
		warn(), error(), err(), critical(), crit(), alert(), emergency() and emerg(), just in case any
		of them are needed.
		I should not have used that design, because it forces other code to provide too much complexity.
	- Update licence to perl_5. See https://metacpan.org/module/CPAN::Meta::Spec#license.
	- Add a note to the POD for header_type(), about the HTTP status being set to 200 when the header type
		has been set to 'none'.
	- Add a note to the FAQ about explicitly outputting a header when using Ajax (at least with jQuery).

1.03  Sat Nov 17 10:31:00 2012
	- Use File::Spec -> catfile('t', 'x.pl') instead of 't/x.pl' in tests.
		This should solve RT#81241 reported by CHORNY under MS Windows.

1.02  Mon Sep 17 11:38:00 2012
	- Add CGI::PSGI to the pre-reqs (Andreas Koenig).
	- Change (in Build.PL) configure_requires{Module::Build} from 0.3800 to 0, after adapting advice from
		 Andreas Koenig.
	- Roll the source of CGI::Snapp::Plugin::Forward into CGI::Snapp.
		See also t/forward.t and t/lib/CGI/Snapp/ForwardTest.pm.
		Update docs.
	- Roll the source of CGI::Snapp::Plugin::Redirect into CGI::Snapp.
		See also t/redirect.t and t/lib/CGI/Snapp/RedirectTest.pm.
		Update docs.

1.01  Sun Apr  8 10:40:00 2012
	- Shift some code from run() into _determine_(cgi|psgi)_output().
	- Add PSGI-specific code, sub as_psgi(), and make some changes to run() using $self -> _psgi().
	- Add t/psgi.basic.pl.
	- Also test via CGI::Snapp::Dispatch's t/psgi.args.t.

1.00  Mon Feb 27 10:09:44 2012
	- Original version


t/isa.pl
use strict;
use warnings;

use CGI::Snapp;

use Test::More;

# ------------------------------------------------

my($count) = 0;

isa_ok(CGI::Snapp -> new(send_output => 0), 'CGI::Snapp'); $count++;

done_testing($count);


t/basic.pl
use strict;
use warnings;

use CGI;
use CGI::Snapp;

use Log::Handler;

use Test::More tests => 4;

# ------------------------------------------------

my($logger) = Log::Handler -> new;

$logger -> add
	(
	 screen =>
	 {
		 maxlevel       => 'debug',
		 message_layout => '%m',
		 minlevel       => 'error',
		 newline        => 1, # When running from the command line.
	 }
	);

my($app) = CGI::Snapp -> new(logger => $logger, QUERY => CGI -> new, send_output => 0);

isa_ok($app, 'CGI::Snapp');

my($modes) = {finish => 'finisher', starter => 'starter'};

$app -> query -> param(rm => 'start');
$app -> run_modes($modes);

is_deeply({$app -> run_modes}, {%$modes, start => 'dump_html'}, 'Set/get run modes');

my($output) = $app -> run;

ok(length($output) > 0, "Output from $0 is not empty");
ok($app -> get_current_runmode eq 'start', "Current run mode is 'start'");


t/params.pl
use lib 't/lib';
use strict;
use warnings;

use CGI::Snapp;

use Log::Handler;

use Test::More;

# ------------------------------------------------

my($logger) = Log::Handler -> new;

$logger -> add
	(
	 screen =>
	 {
		 maxlevel       => 'debug',
		 message_layout => '%m',
		 minlevel       => 'error',
		 newline        => 1, # When running from the command line.
	 }
	);

my($count) = 0;
my($app)   = CGI::Snapp -> new(logger => $logger, send_output => 0);

# Get the default params, of which there are none.

is_deeply([$app -> param], [], 'No params are set by default'); $count++;

# Set/get a hash of params.

my(%old_params) = (one => 1, two => 2);

$app -> param(%old_params);

my(%new_params) = map{($_ => $app -> param($_) )} $app -> param;

is_deeply(\%old_params, \%new_params, 'Can set and get a hash of params'); $count++;

$app -> delete($_) for keys %old_params;

is_deeply([$app -> param], [], 'No params are set after mass delete'); $count++;

# Set/get a hashref of params.

my($old_params) = {one => 1, two => 2};

$app -> param($old_params);

%new_params = map{($_ => $app -> param($_) ) } $app -> param;

is_deeply($old_params, \%new_params, 'Can set and get a hash of params'); $count++;

$app -> delete($_) for keys %$old_params;

is_deeply([$app -> param], [], 'No params are set after mass delete'); $count++;

# Add params in stages.

%old_params = (one => 1, two => 2);

$app -> param(%old_params);

%old_params = (three => 3, four => 4);

$app -> param(%old_params);

%old_params = (five => 5);

my($value) = $app -> param(%old_params);

ok($value == 5, 'param($key => $value) returns that value');                $count++;
ok($value == $app -> param('five'), 'param($key) also returns that value'); $count++;

%old_params = (one => 1, two => 2, three => 3, four => 4, %old_params);

%new_params = map{($_ => $app -> param($_) )} $app -> param;

is_deeply(\%old_params, \%new_params, 'Params match after being added in stages'); $count++;

# What is returned after delete()?

my($key) = 'five';
$value   = $app -> delete($key);

ok($value == 5, "delete($key) returns the corresponding value: $value");            $count++;
ok(! defined $app -> param($key), "After delete($key), param($key) returns undef"); $count++;

$value = $app -> delete($key);

ok(! defined $value, "After delete($key) twice, delete() returns undef"); $count++;

# Clean up before testing an arrayref.

$app -> delete($_) for keys %old_params;

$old_params = [qw/six 6 seven 7/];

$app -> param($old_params);

%new_params = map{($_ => $app -> param($_) )} $app -> param;

is_deeply($old_params, [%new_params], 'Params match after being added via an arrayref'); $count++;

done_testing($count);


Build.PL
use Module::Build;

Module::Build -> new
(
	module_name    => 'CGI::Snapp',
	license        => 'artistic_2',
	dist_abstract  => 'An almost back-compat fork of CGI::Application',
	dist_author    => 'Ron Savage <ron@savage.net.au>',
	build_requires =>
	{
		Test::More=> 0.94,
#		Test::Pod  => 1.45, # Make it optional. See t/pod.t
	},
	configure_requires =>
	{
		Module::Build => 0,
	},
	requires =>
	{
		Capture::Tiny   => 0.16,
		Carp            => 1.20,
		CGI             => 3.59,
		CGI::PSGI       => 0.15,
		CGI::Simple     => 1.113,
		Class::ISA      => 0.36,
		File::Spec      => 0.86,
		Hash::FieldHash => 0.10,
		IO::Pipe        => 1.13,
		Log::Handler    => 0.75,
		parent          => 0.225,
		Proc::Fork      => 0.71,
		strict          => 1.03,
		Try::Tiny       => 0.11,
		vars            => 1.01,
		warnings        => 1.03,
	},
) -> create_build_script();


Makefile.PL
use ExtUtils::MakeMaker;
	LICENSE   => 'artistic_2',

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile
(
	($] ge '5.005') ?
	(
	 AUTHOR   => 'Ron Savage (ron@savage.net.au)',
	 ABSTRACT => 'An almost back-compat fork of CGI::Application',
	   ) : (),
	clean =>
	{
		FILES => 'blib/* Makefile MANIFEST CGI-Snapp-*'
	},
	dist =>
	{
		COMPRESS => 'gzip',
		SUFFIX   => 'gz'
	},
	DISTNAME  => 'CGI-Snapp',
	NAME      => 'CGI::Snapp',
	LICENSE   => 'artistic_2',
	PL_FILES  => {},
	PREREQ_PM =>
	{
		Capture::Tiny   => 0.16,
		Carp            => 1.20,
		CGI             => 3.59,
		CGI::PSGI       => 0.15,
		CGI::Simple     => 1.113,
		Class::ISA      => 0.36,
		File::Spec      => 0.86,
		Hash::FieldHash => 0.10,
		IO::Pipe        => 1.13,
		Log::Handler    => 0.75,
		parent          => 0.225,
		Proc::Fork      => 0.71,
		strict          => 1.03,
		Test::More    => 0.94,
#		Test::Pod       => 1.45, # Make it optional. See t/pod.t
		Try::Tiny       => 0.11,
		vars            => 1.01,
		warnings        => 1.03,
	},
	VERSION_FROM => 'lib/CGI/Snapp.pm',
);
